File: app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/RerouteActivity.java
package com.mapbox.services.android.navigation.testapp.activity;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.view.View;
import android.widget.Toast;

import com.mapbox.android.core.location.LocationEngineCallback;
import com.mapbox.android.core.location.LocationEngineResult;
import com.mapbox.api.directions.v5.models.DirectionsResponse;
import com.mapbox.api.directions.v5.models.DirectionsRoute;
import com.mapbox.core.constants.Constants;
import com.mapbox.geojson.LineString;
import com.mapbox.geojson.Point;
import com.mapbox.mapboxsdk.Mapbox;
import com.mapbox.mapboxsdk.annotations.MarkerOptions;
import com.mapbox.mapboxsdk.annotations.Polyline;
import com.mapbox.mapboxsdk.annotations.PolylineOptions;
import com.mapbox.mapboxsdk.camera.CameraPosition;
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory;
import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.location.LocationComponent;
import com.mapbox.mapboxsdk.location.modes.RenderMode;
import com.mapbox.mapboxsdk.maps.MapView;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
import com.mapbox.mapboxsdk.maps.Style;
import com.mapbox.services.android.navigation.testapp.R;
import com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView;
import com.mapbox.services.android.navigation.v5.location.replay.ReplayRouteLocationEngine;
import com.mapbox.services.android.navigation.v5.milestone.Milestone;
import com.mapbox.services.android.navigation.v5.milestone.MilestoneEventListener;
import com.mapbox.services.android.navigation.v5.milestone.VoiceInstructionMilestone;
import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation;
import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigationOptions;
import com.mapbox.services.android.navigation.v5.navigation.NavigationEventListener;
import com.mapbox.services.android.navigation.v5.navigation.NavigationRoute;
import com.mapbox.services.android.navigation.v5.offroute.OffRouteListener;
import com.mapbox.services.android.navigation.v5.routeprogress.ProgressChangeListener;
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import timber.log.Timber;

public class RerouteActivity extends HistoryActivity implements OnMapReadyCallback,
  Callback<DirectionsResponse>, MapboxMap.OnMapClickListener, NavigationEventListener,
  OffRouteListener, ProgressChangeListener, MilestoneEventListener {

  @BindView(R.id.mapView)
  MapView mapView;
  @BindView(android.R.id.content)
  View contentLayout;
  @BindView(R.id.instructionView)
  InstructionView instructionView;

  private Point origin = Point.fromLngLat(-0.358764, 39.494876);
  private Point destination = Point.fromLngLat(-0.383524, 39.497825);
  private Polyline polyline;

  private final RerouteActivityLocationCallback callback = new RerouteActivityLocationCallback(this);
  private Location lastLocation;
  private ReplayRouteLocationEngine mockLocationEngine;
  private MapboxNavigation navigation;
  private MapboxMap mapboxMap;
  private boolean running;
  private boolean tracking;
  private boolean wasInTunnel = false;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    setTheme(R.style.NavigationViewLight);
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_reroute);
    ButterKnife.bind(this);

    mapView.onCreate(savedInstanceState);
    mapView.getMapAsync(this);

    MapboxNavigationOptions options = MapboxNavigationOptions.builder().isDebugLoggingEnabled(true).build();
    navigation = new MapboxNavigation(getApplicationContext(), Mapbox.getAccessToken(), options);
    navigation.addNavigationEventListener(this);
    navigation.addMilestoneEventListener(this);
    addNavigationForHistory(navigation);

    instructionView.retrieveSoundButton().show();
    instructionView.retrieveSoundButton().addOnClickListener(
      v -> Toast.makeText(RerouteActivity.this, "Sound button clicked!", Toast.LENGTH_SHORT).show()
    );
  }

  @Override
  public void onResume() {
    super.onResume();
    mapView.onResume();
  }

  @Override
  protected void onStart() {
    super.onStart();
    mapView.onStart();
  }

  @Override
  protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    mapView.onSaveInstanceState(outState);
  }

  @Override
  protected void onStop() {
    super.onStop();
    mapView.onStop();
  }

  @Override
  public void onPause() {
    super.onPause();
    mapView.onPause();
  }

  @Override
  public void onLowMemory() {
    super.onLowMemory();
    mapView.onLowMemory();
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    mapView.onDestroy();
    shutdownLocationEngine();
    shutdownNavigation();
  }

  @SuppressLint("MissingPermission")
  @Override
  public void onMapReady(@NonNull MapboxMap mapboxMap) {
    this.mapboxMap = mapboxMap;
    this.mapboxMap.addOnMapClickListener(this);
    mapboxMap.setStyle(Style.DARK, style -> {
      LocationComponent locationComponent = mapboxMap.getLocationComponent();
      locationComponent.activateLocationComponent(this, style);
      locationComponent.setLocationComponentEnabled(true);
      locationComponent.setRenderMode(RenderMode.GPS);

      mockLocationEngine = new ReplayRouteLocationEngine();
      getRoute(origin, destination);
    });
  }

  @Override
  public boolean onMapClick(@NonNull LatLng point) {
    if (!running || mapboxMap == null || lastLocation == null) {
      return true;
    }

    mapboxMap.addMarker(new MarkerOptions().position(point));
    mapboxMap.removeOnMapClickListener(this);

    destination = Point.fromLngLat(point.getLongitude(), point.getLatitude());
    resetLocationEngine(destination);

    tracking = false;
    return true;
  }

  @Override
  public void onRunning(boolean running) {
    this.running = running;
    if (running) {
      navigation.addOffRouteListener(this);
      navigation.addProgressChangeListener(this);
    }
  }

  @Override
  public void userOffRoute(Location location) {
    origin = Point.fromLngLat(lastLocation.getLongitude(), lastLocation.getLatitude());
    getRoute(origin, destination);
    Snackbar.make(contentLayout, "User Off Route", Snackbar.LENGTH_SHORT).show();
    mapboxMap.addMarker(new MarkerOptions().position(new LatLng(location.getLatitude(), location.getLongitude())));
  }

  @Override
  public void onProgressChange(Location location, RouteProgress routeProgress) {
    boolean isInTunnel = routeProgress.inTunnel();
    lastLocation = location;
    if (!wasInTunnel && isInTunnel) {
      wasInTunnel = true;
      Snackbar.make(contentLayout, "Enter tunnel!", Snackbar.LENGTH_SHORT).show();
    }
    if (wasInTunnel && !isInTunnel) {
      wasInTunnel = false;
      Snackbar.make(contentLayout, "Exit tunnel!", Snackbar.LENGTH_SHORT).show();
    }
    if (tracking) {
      mapboxMap.getLocationComponent().forceLocationUpdate(location);
      CameraPosition cameraPosition = new CameraPosition.Builder()
        .zoom(15)
        .target(new LatLng(location.getLatitude(), location.getLongitude()))
        .bearing(location.getBearing())
        .build();
      mapboxMap.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition), 2000);
    }
    instructionView.updateDistanceWith(routeProgress);
  }

  @Override
  public void onMilestoneEvent(RouteProgress routeProgress, String instruction, Milestone milestone) {
    if (milestone instanceof VoiceInstructionMilestone) {
      Snackbar.make(contentLayout, instruction, Snackbar.LENGTH_SHORT).show();
    }
    instructionView.updateBannerInstructionsWith(milestone);
    Timber.d("onMilestoneEvent - Current Instruction: %s", instruction);
  }

  @Override
  public void onResponse(@NonNull Call<DirectionsResponse> call, @NonNull Response<DirectionsResponse> response) {
    Timber.d(call.request().url().toString());
    if (response.body() != null) {
      if (!response.body().routes().isEmpty()) {
        DirectionsRoute route = response.body().routes().get(0);
        drawRoute(route);
        resetLocationEngine(route);
        navigation.startNavigation(route);
        mapboxMap.addOnMapClickListener(this);
        tracking = true;
      }
    }
  }

  @Override
  public void onFailure(@NonNull Call<DirectionsResponse> call, @NonNull Throwable throwable) {
    Timber.e(throwable);
  }

  void updateLocation(Location location) {
    if (!tracking) {
      mapboxMap.getLocationComponent().forceLocationUpdate(location);
    }
  }

  private void getRoute(Point origin, Point destination) {
    NavigationRoute.builder(this)
      .origin(origin)
      .destination(destination)
      .accessToken(Mapbox.getAccessToken())
      .build().getRoute(this);
  }

  private void drawRoute(DirectionsRoute route) {
    List<LatLng> points = new ArrayList<>();
    List<Point> coords = LineString.fromPolyline(route.geometry(), Constants.PRECISION_6).coordinates();

    for (Point point : coords) {
      points.add(new LatLng(point.latitude(), point.longitude()));
    }

    if (!points.isEmpty()) {
      if (polyline != null) {
        mapboxMap.removePolyline(polyline);
      }
      polyline = mapboxMap.addPolyline(new PolylineOptions()
        .addAll(points)
        .color(Color.parseColor(getString(R.string.blue)))
        .width(5));
    }
  }

  private void resetLocationEngine(Point point) {
    mockLocationEngine.moveTo(point);
    navigation.setLocationEngine(mockLocationEngine);
  }

  private void resetLocationEngine(DirectionsRoute directionsRoute) {
    mockLocationEngine.assign(directionsRoute);
    navigation.setLocationEngine(mockLocationEngine);
  }

  private void shutdownLocationEngine() {
    if (mockLocationEngine != null) {
      mockLocationEngine.removeLocationUpdates(callback);
    }
  }

  private void shutdownNavigation() {
    navigation.removeNavigationEventListener(this);
    navigation.removeProgressChangeListener(this);
    navigation.onDestroy();
  }

  private static class RerouteActivityLocationCallback implements LocationEngineCallback<LocationEngineResult> {

    private final WeakReference<RerouteActivity> activityWeakReference;

    RerouteActivityLocationCallback(RerouteActivity activity) {
      this.activityWeakReference = new WeakReference<>(activity);
    }

    @Override
    public void onSuccess(LocationEngineResult result) {
      RerouteActivity activity = activityWeakReference.get();
      if (activity != null) {
        Location location = result.getLastLocation();
        if (location == null) {
          return;
        }
        activity.updateLocation(location);
      }
    }

    @Override
    public void onFailure(@NonNull Exception exception) {
      Timber.e(exception);
    }
  }
}

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReRouteActivity.kt
package com.mapbox.navigation.examples.core

import android.annotation.SuppressLint
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.Snackbar.LENGTH_SHORT
import com.mapbox.android.core.location.LocationEngineCallback
import com.mapbox.android.core.location.LocationEngineRequest
import com.mapbox.android.core.location.LocationEngineResult
import com.mapbox.api.directions.v5.DirectionsCriteria
import com.mapbox.api.directions.v5.MapboxDirections
import com.mapbox.api.directions.v5.models.DirectionsResponse
import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.api.directions.v5.models.RouteOptions
import com.mapbox.geojson.Point
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory
import com.mapbox.mapboxsdk.location.modes.RenderMode
import com.mapbox.mapboxsdk.maps.MapboxMap
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback
import com.mapbox.mapboxsdk.maps.Style
import com.mapbox.navigation.base.internal.extensions.applyDefaultParams
import com.mapbox.navigation.base.internal.extensions.coordinates
import com.mapbox.navigation.base.trip.model.RouteProgress
import com.mapbox.navigation.core.MapboxNavigation
import com.mapbox.navigation.core.directions.session.RoutesObserver
import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
import com.mapbox.navigation.core.reroute.RerouteController
import com.mapbox.navigation.core.reroute.RerouteState
import com.mapbox.navigation.core.trip.session.LocationObserver
import com.mapbox.navigation.core.trip.session.OffRouteObserver
import com.mapbox.navigation.core.trip.session.RouteProgressObserver
import com.mapbox.navigation.core.trip.session.TripSessionState
import com.mapbox.navigation.core.trip.session.TripSessionStateObserver
import com.mapbox.navigation.examples.R
import com.mapbox.navigation.examples.utils.Utils
import com.mapbox.navigation.examples.utils.extensions.toPoint
import com.mapbox.navigation.ui.camera.NavigationCamera
import com.mapbox.navigation.ui.map.NavigationMapboxMap
import java.lang.ref.WeakReference
import kotlinx.android.synthetic.main.activity_reroute_layout.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber

/**
 * This activity shows how to:
 * - observe reroute events with the Navigation SDK's [RoutesObserver];
 * - replace default [RerouteController] and handle re-route events.
 */
class ReRouteActivity : AppCompatActivity(), OnMapReadyCallback, OffRouteObserver,
    RerouteController.RerouteStateObserver {

    companion object {
        private const val DEFAULT_FASTEST_INTERVAL = 500L
        private const val DEFAULT_ENGINE_REQUEST_INTERVAL = 1000L
    }

    private var mapboxNavigation: MapboxNavigation? = null
    private var navigationMapboxMap: NavigationMapboxMap? = null
    private var directionRoute: DirectionsRoute? = null
    private val routeSettings = CoordinatesHolder()

    private val locationListenerCallback = MyLocationEngineCallback(this)
    private val tripSessionStateObserver = object : TripSessionStateObserver {
        override fun onSessionStateChanged(tripSessionState: TripSessionState) {
            when (tripSessionState) {
                TripSessionState.STARTED -> {
                    startNavigation.visibility = GONE
                    stopLocationUpdates()
                    mapboxNavigation?.registerRouteProgressObserver(routeProgressObserver)
                }
                TripSessionState.STOPPED -> {
                    startLocationUpdates()
                    navigationMapboxMap?.hideRoute()
                    updateCameraOnNavigationStateChange(false)
                }
            }
        }
    }

    private val routeObserver = object : RoutesObserver {
        override fun onRoutesChanged(routes: List<DirectionsRoute>) {
            startNavigation.visibility = if (routes.isNotEmpty()) {
                VISIBLE
            } else {
                GONE
            }
            directionRoute = routes.firstOrNull()
            navigationMapboxMap?.drawRoutes(routes)
        }
    }
    private val routeProgressObserver = object : RouteProgressObserver {
        override fun onRouteProgressChanged(routeProgress: RouteProgress) {
        }
    }
    private val routesReqCallback = object : RoutesRequestCallback {
        override fun onRoutesReady(routes: List<DirectionsRoute>) {
            Timber.d("route request success %s", routes.toString())
        }

        override fun onRoutesRequestFailure(throwable: Throwable, routeOptions: RouteOptions) {
            Timber.e("route request failure %s", throwable.toString())
        }

        override fun onRoutesRequestCanceled(routeOptions: RouteOptions) {
            Timber.d("route request canceled")
        }
    }

    private val locationObserver = object : LocationObserver {
        override fun onRawLocationChanged(rawLocation: Location) {
            routeSettings.currentLocation = rawLocation
        }

        override fun onEnhancedLocationChanged(
            enhancedLocation: Location,
            keyPoints: List<Location>
        ) = Unit
    }

    @SuppressLint("MissingPermission")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_reroute_layout)
        mapView.onCreate(savedInstanceState)
        mapView.getMapAsync(this)

        val mapboxNavigationOptions = MapboxNavigation
            .defaultNavigationOptionsBuilder(this, Utils.getMapboxAccessToken(this))
            .build()

        mapboxNavigation = MapboxNavigation(mapboxNavigationOptions)
        mapboxNavigation?.setRerouteController(
            SampleRerouteController(Utils.getMapboxAccessToken(this), routeSettings)
        )
        initListeners()
    }

    override fun onStart() {
        super.onStart()
        mapView.onStart()

        mapboxNavigation?.run {
            registerLocationObserver(locationObserver)
            registerTripSessionStateObserver(tripSessionStateObserver)
            registerRoutesObserver(routeObserver)
            registerRouteProgressObserver(routeProgressObserver)
            registerOffRouteObserver(this@ReRouteActivity)
            getRerouteController()?.registerRerouteStateObserver(this@ReRouteActivity)
        }
    }

    public override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    public override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

    override fun onStop() {
        super.onStop()
        mapboxNavigation?.run {
            unregisterLocationObserver(locationObserver)
            unregisterTripSessionStateObserver(tripSessionStateObserver)
            unregisterRouteProgressObserver(routeProgressObserver)
            unregisterRoutesObserver(routeObserver)
            unregisterOffRouteObserver(this@ReRouteActivity)
            getRerouteController()?.unregisterRerouteStateObserver(this@ReRouteActivity)
        }
        stopLocationUpdates()
        mapView.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        mapboxNavigation?.stopTripSession()
        mapboxNavigation?.onDestroy()
        mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView.onLowMemory()
    }

    @SuppressLint("MissingPermission")
    override fun onMapReady(mapboxMap: MapboxMap) {
        mapboxMap.setStyle(Style.MAPBOX_STREETS) { style ->
            mapboxMap.moveCamera(CameraUpdateFactory.zoomTo(15.0))
            navigationMapboxMap = NavigationMapboxMap(mapView, mapboxMap, this, true)

            mapboxNavigation?.navigationOptions?.locationEngine?.getLastLocation(
                locationListenerCallback
            )

            directionRoute?.let {
                navigationMapboxMap?.drawRoute(it)
                mapboxNavigation?.setRoutes(listOf(it))
                startNavigation.visibility = View.VISIBLE
            }
        }

        mapboxMap.addOnMapLongClickListener { latLng ->
            routeSettings.destination = latLng.toPoint()
            mapboxMap.locationComponent.lastKnownLocation?.let { originLocation ->
                mapboxNavigation?.requestRoutes(
                    RouteOptions.builder().applyDefaultParams()
                        .accessToken(Utils.getMapboxAccessToken(applicationContext))
                        .coordinates(originLocation.toPoint(), null, latLng.toPoint())
                        .alternatives(true)
                        .profile(DirectionsCriteria.PROFILE_DRIVING_TRAFFIC)
                        .build(),
                    routesReqCallback
                )
            }
            true
        }

        if (directionRoute == null) {
            Snackbar.make(container, R.string.msg_long_press_map_to_place_waypoint, LENGTH_SHORT)
                .show()
        }
    }

    @SuppressLint("MissingPermission")
    private fun initListeners() {
        startNavigation.setOnClickListener {
            updateCameraOnNavigationStateChange(true)
            navigationMapboxMap?.addProgressChangeListener(mapboxNavigation!!)
            if (mapboxNavigation?.getRoutes()?.isNotEmpty() == true) {
                navigationMapboxMap?.startCamera(mapboxNavigation?.getRoutes()!![0])
            }
            mapboxNavigation?.startTripSession()
            startNavigation.visibility = View.GONE
            stopLocationUpdates()
        }
    }

    @SuppressLint("MissingPermission")
    private fun startLocationUpdates() {
        val requestLocationUpdateRequest =
            LocationEngineRequest.Builder(DEFAULT_ENGINE_REQUEST_INTERVAL)
                .setFastestInterval(DEFAULT_FASTEST_INTERVAL)
                .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
                .build()

        mapboxNavigation?.navigationOptions?.locationEngine?.requestLocationUpdates(
            requestLocationUpdateRequest,
            locationListenerCallback,
            mainLooper
        )
    }

    private fun stopLocationUpdates() {
        mapboxNavigation?.navigationOptions?.locationEngine?.removeLocationUpdates(
            locationListenerCallback
        )
    }

    private fun updateCameraOnNavigationStateChange(
        navigationStarted: Boolean
    ) {
        navigationMapboxMap?.apply {
            if (navigationStarted) {
                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_GPS)
                updateLocationLayerRenderMode(RenderMode.GPS)
            } else {
                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_NONE)
                updateLocationLayerRenderMode(RenderMode.COMPASS)
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        // This is not the most efficient way to preserve the route on a device rotation.
        // This is here to demonstrate that this event needs to be handled in order to
        // redraw the route line after a rotation.
        directionRoute?.let {
            outState.putString(Utils.PRIMARY_ROUTE_BUNDLE_KEY, it.toJson())
        }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
        super.onRestoreInstanceState(savedInstanceState)
        directionRoute = Utils.getRouteFromBundle(savedInstanceState)
    }

    override fun onOffRouteStateChanged(offRoute: Boolean) {
        if (offRoute) {
            Toast.makeText(this, "You're off-route", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onRerouteStateChanged(rerouteState: RerouteState) {
        val message: String = when (rerouteState) {
            RerouteState.Idle -> "Reroute is idle"
            RerouteState.Interrupted -> "Reroute request interrupted"
            is RerouteState.Failed -> "Reroute request failed: ${rerouteState.message}"
            RerouteState.FetchingRoute -> "Reroute request is in progress"
            RerouteState.RouteFetched -> "Reroute request is finished successful"
        }
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    private class MyLocationEngineCallback(activity: ReRouteActivity) :
        LocationEngineCallback<LocationEngineResult> {

        private val activityRef = WeakReference(activity)

        override fun onSuccess(result: LocationEngineResult?) {
            result?.locations?.firstOrNull()?.let {
                val activity = activityRef.get() ?: return@let
                activity.navigationMapboxMap?.updateLocation(result.lastLocation)
            }
        }

        override fun onFailure(exception: Exception) {
        }
    }

    private class SampleRerouteController(
        private val accessToken: String,
        private val coordinatesHolder: CoordinatesHolder
    ) : RerouteController {
        private val stateObservers = mutableSetOf<RerouteController.RerouteStateObserver>()

        private var mapboxDirections: MapboxDirections? = null
        private val mainHandler = Handler(Looper.getMainLooper())

        override var state: RerouteState = RerouteState.Idle
            set(value) {
                if (field == value) {
                    return
                }
                field = value
                stateObservers.forEach { it.onRerouteStateChanged(field) }
            }

        override fun reroute(routesCallback: RerouteController.RoutesCallback) {
            state = RerouteState.FetchingRoute
            mapboxDirections = MapboxDirections.builder()
                .accessToken(accessToken)
                .profile(DirectionsCriteria.PROFILE_DRIVING_TRAFFIC)
                .origin(coordinatesHolder.currentLocation.toPoint())
                .destination(coordinatesHolder.destination)
                .steps(true)
                .build()

            mapboxDirections?.enqueueCall(object : Callback<DirectionsResponse> {
                override fun onResponse(
                    call: Call<DirectionsResponse>,
                    response: Response<DirectionsResponse>
                ) {
                    mainHandler.post {
                        val routes = response.body()?.routes()
                        when {
                            call.isCanceled -> {
                                state = RerouteState.Interrupted
                            }
                            response.isSuccessful && !routes.isNullOrEmpty() -> {
                                state = RerouteState.RouteFetched
                                routesCallback.onNewRoutes(routes)
                            }
                            else -> {
                                state = RerouteState.Failed("Reroute request is empty")
                            }
                        }
                        state = RerouteState.Idle
                    }
                }

                override fun onFailure(call: Call<DirectionsResponse>, t: Throwable) {
                    mainHandler.post {
                        state = if (call.isCanceled) {
                            RerouteState.Interrupted
                        } else {
                            RerouteState.Failed("Reroute request failed", t)
                        }
                        state = RerouteState.Idle
                    }
                }
            })
        }

        override fun interrupt() {
            if (state == RerouteState.FetchingRoute) {
                mapboxDirections?.cancelCall()
                mapboxDirections = null
            }
        }

        override fun registerRerouteStateObserver(rerouteStateObserver: RerouteController.RerouteStateObserver): Boolean {
            rerouteStateObserver.onRerouteStateChanged(state)
            return stateObservers.add(rerouteStateObserver)
        }

        override fun unregisterRerouteStateObserver(rerouteStateObserver: RerouteController.RerouteStateObserver): Boolean {
            return stateObservers.remove(rerouteStateObserver)
        }
    }

    private class CoordinatesHolder {
        var currentLocation: Location = Location("InternalProvider")
        var destination: Point = Point.fromLngLat(0.0, 0.0)
    }
}
